### PR TITLE
[lua] get image from database by image id

### PR DESCRIPTION
**Pull request by wpferguson**
Added a get_image function to darktable.database to retrieve the image by imageid.

The only lua function for retrieving an image from the database is darktable.database[index].  This index specifies the row of the image table, not the image id.  As long as no images are removed or trashed, the index and the row id agree, but once an import is done, removed, then reimported the row id differs from the image id by the amount of images removed.  Therefore, the only way to retrieve a specific image is by searching the database.  

Attached is a test script that retrieves the last row of the image table, then retrieves the same image by image id.  If any deleting or removing of images has occurred, then the two numbers should not agree, which demonstrates the problem and the solution.
[test_db_get_image.lua.txt](https://github.com/darktable-org/darktable/files/5839449/test_db_get_image.lua.txt)

### CODE REVIEW

File: src/lua/database.c [hunk 1]
```diff
@@ -206,6 +206,32 @@ static int database_numindex(lua_State *L)
...
+static int database_get_image(lua_State *L)
+{
+  int img_id = luaL_checkinteger(L, -1);
+  if(img_id < 1)
+  {
+    return luaL_error(L, "incorrect image id in database");
+  }
+  sqlite3_stmt *stmt = NULL;
+  char query[1024];
+  snprintf(query, sizeof(query), "SELECT id FROM main.images WHERE id = %d LIMIT 1",
```
Comment (TurboGit) [inline src/lua/database.c:218]: Just to be sure, this returns img_id (the parameter) if image is found. So this routine is basically equivalent to a check if an image id exists in the database. Right ?

File: src/lua/database.c [hunk 2]
```diff
@@ -206,6 +206,32 @@ static int database_numindex(lua_State *L)
   return 1;
 }
 
+static int database_get_image(lua_State *L)
+{
+  int img_id = luaL_checkinteger(L, -1);
```
Comment (TurboGit) [inline src/lua/database.c:211]: const

**Review by TurboGit — COMMENTED** [review]
Just a comment/question to be sure I understand the purpose. TIA.

File: src/lua/database.c [hunk 3]
```diff
@@ -206,6 +206,32 @@ static int database_numindex(lua_State *L)
...
+  }
+  sqlite3_stmt *stmt = NULL;
+  char query[1024];
+  snprintf(query, sizeof(query), "SELECT id FROM main.images WHERE id = %d LIMIT 1",
+           img_id);
+  DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db), query, -1, &stmt, NULL);
+  if(sqlite3_step(stmt) == SQLITE_ROW)
+  {
+    int imgid = sqlite3_column_int(stmt, 0);
+    luaA_push(L, dt_lua_image_t, &imgid);
```
Comment (wpferguson) [inline src/lua/database.c:224]: Here we push the pointer to the image

File: src/lua/database.c (hunk 1 above)
Comment (wpferguson) [reply src/lua/database.c:218]: It certainly looks that way :), but the return value from the query is the image object, so imgid != img_id

File: src/lua/database.c (hunk 2 above)
Comment (wpferguson) [reply src/lua/database.c]: done

File: src/lua/database.c (hunk 1 above)
Comment (TurboGit) [reply src/lua/database.c:218]: I see, lua confused me :)

File: src/lua/database.c [hunk 4]
```diff
@@ -206,6 +206,32 @@ static int database_numindex(lua_State *L)
...
+    sqlite3_finalize(stmt);
+  }
+  else
+  {
+    sqlite3_finalize(stmt);
+    lua_pushnil(L);
+  }
+  return 1;
```
Comment (TurboGit) [inline src/lua/database.c:232]: Why not a single `sqlite3_finalize(stmt);` after the if/else ?

Comment (wpferguson) [reply src/lua/database.c:232]: because I copied from above and just changed enough to make it work, instead of proof reading :).  I'll fix them both

**Review by TurboGit — APPROVED** [review]
All good now, thanks!

**Pull request merged**